File: in_Java/0207 Course Schedule.java
class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        int[][] matrix = new int[numCourses][numCourses];
        int[] incoming_degree = new int[numCourses];
        for(int i = 0; i < prerequisites.length; i++){
            int pre = prerequisites[i][1];
            int cur = prerequisites[i][0];
            if(matrix[pre][cur] == 0) incoming_degree[cur]++;
            matrix[pre][cur] = 1;
        }
        Queue<Integer> q = new LinkedList<>();
        for(int i = 0; i < numCourses; i++){
            if(incoming_degree[i] == 0) q.add(i);
        }
        int count = numCourses;
        // now queue contains nodes with no incoming edges
        while(q.size()!= 0){
            int now = q.poll();
            count--;
            for(int i = 0; i < numCourses; i++){
                if(matrix[now][i] != 0){
                    incoming_degree[i]--;
                    if(incoming_degree[i] == 0) q.add(i);
                }
            }
        }
        return count == 0;
    }
}

File: in_Python/0207 Course Schedule.py
class Solution(object):
    def canFinish(self, numCourses, prerequisites):
        """
        :type numCourses: int
        :type prerequisites: List[List[int]]
        :rtype: bool
        """
        coursemap = {}
        degree = [0] * numCourses
        for i in range(numCourses):
            coursemap[i] = []
        for p in prerequisites:
            t = p[0]
            s = p[1]
            coursemap[s].append(t)
            degree[t] += 1
        res = []
        q = []
        for i in range(numCourses):
            if degree[i] == 0:
                q.append(i)
        for i in q:
            for j in coursemap[i]:
                degree[j] -= 1
                if degree[j] == 0:
                    q.append(j)
        return len(q) == numCourses
        
        
        